[KMSv2] add docs for KMSv2 GA (#43398)
kubernetes/website: content/en/docs/tasks/administer-cluster/kms-provider.md

Reviewer: It'd be nice to revise the complex “Notes” section near the end of this page.

<!-- overview -->
This page shows how to configure a Key Management Service (KMS) provider and plugin to enable secret data encryption.
In Kubernetes {{< skew currentVersion >}} there are two versions of KMS at-rest encryption.
You should use KMS v2 if feasible because KMS v1 is deprecated (since Kubernetes v1.28) and disabled by default (since Kubernetes v1.29).

> However, you should also read and observe the **Caution** notices in this page that highlight specific
> cases when you must not use KMS v2.  KMS v2 offers significantly better performance characteristics than KMS v1.

Reviewer 2: ```suggestion
KMS v2 offers significantly better performance characteristics than KMS v1.
```

> 

Reviewer: ```suggestion

{{< caution >}}
This documentation is for the generally available implementations of KMS v1 and v2.
If you are using any control plane components older than Kubernetes v1.29, please check
the equivalent page in the documentation for the version of Kubernetes that your cluster
is running.
{{< /caution >}}

```
?

Author: @sftim I have removed all the references to old Kubernetes version from the doc, so wasn't sure if this is required. LMK what you think!

> This documentation is for the generally available implementations of KMS v1 and v2.

KMS v1 was in beta and has been deprecated. Only KMS v2 is GA.

Reviewer: Try this, then:
```suggestion

{{< caution >}}
This documentation is for the generally available implementation of KMS v2 (and for the
deprecated version 1 implementation).
If you are using any control plane components older than Kubernetes v1.29, please check
the equivalent page in the documentation for the version of Kubernetes that your cluster
is running. Earlier releases on Kubernetes had different behavior that may be relevant
for information security.
{{< /caution >}}

```
?

Author: added it!

* Kubernetes version 1.10.0 or later is required

* For version 1.29 and later, the v1 implementation of KMS is disabled by default.
  To enable the feature, set `--feature-gates=KMSv1=true` to configure a KMS v1 provider.

* Your cluster must use etcd v3 or later

### KMS v2
{{< feature-state for_k8s_version="v1.29" state="stable" >}}

* Your cluster must use etcd v3 or later

<!-- steps -->

> 

Reviewer: ```suggestion

## KMS encryption and per-object encryption keys

```
Otherwise, this reads like more of the prerequisites.

* Kubernetes version 1.10.0 or later is required

> * For version 1.29 and later, the feature is disabled by default.

Reviewer: ```suggestion
* For version 1.29 and later, the v1 implementation of KMS is disabled by default.
```

  To enable the feature, set `--feature-gates=KMSv1=true` to configure a KMS v1 provider.

* Your cluster must use etcd v3 or later

### KMS v2
{{< feature-state for_k8s_version="v1.29" state="stable" >}}

> * For version 1.25 and 1.26, enabling the feature via kube-apiserver feature gate is required.

Reviewer: ```suggestion
```
The feature-state shortcode is a hint that the doc won't apply to older versions (or is certainly intended to be).

> Set `--feature-gates=KMSv2=true` to configure a KMS v2 provider.

Reviewer: ```suggestion
Set `--feature-gates=KMSv2=true` to configure a KMS v2 provider.
```

---

`KMSv2KDF` feature gate is enabled by default in v1.29 and cannot be disabled.
Running mixed API server versions with some servers at v1.28 _with the `KMSv2KDF` feature gate disabled_,

> and others at v1.29 is **not supported** - and is likely to result in data loss.

Reviewer 3: > Running mixed API server versions with some servers at v1.27, and others at v1.28 _with the
`KMSv2KDF` feature gate enabled_  is **not supported** - and is likely to result in data loss.

This warning is still true. In general do we expect users to click the version drop down to look at content thats relevant to older releases or should we still include warnings like this even in the latest v1.29+ doc?

> and others at v1.29 is **not supported** - and is likely to result in data loss.

Reviewer: One day I'd like a shortcode to make this easier, but what I'd do for now is something like:
```markdown

{{< note >}}
This documentation is for the stable implementation of KMSv2. If you are using any control plane
components older than Kubernetes v1.29, please check the equivalent page in the documentation
for the release your cluster is running.
{{< /note >}}

```

> and others at v1.29 is **not supported** - and is likely to result in data loss.

Reviewer 2: I commented on slack but there is no potential for data loss between 1.28 and 1.29, so this specific callout is no longer necessary.

> With KMS v1, a new DEK is generated for each encryption.
> 

Reviewer: ```suggestion
If you use the (deprecated) v1 implementation of KMS, a new DEK is generated for each encryption.
```

Reviewer: We might not need to mention the difference, as (AIUI) both v1 and v2 generate a DEK per encryption.

With KMS v2, a new DEK is generated **per encryption**: the API server uses a
_key derivation function_ to generate single use data encryption keys from a secret seed
combined with some random data.
The seed is rotated whenever the KEK is rotated

> (see `Understanding key_id and Key Rotation` section below for more details).

Reviewer: ```suggestion
(see the _Understanding key_id and Key Rotation_ section below for more details).
```
_or_ use a hyperlink, adding `<a id="key-id-and-rotation"></a>` at an appropriate place to create a fragment identifier.

##### KMS v2 {#developing-a-kms-plugin-gRPC-server-notes-kms-v2}

* KMS plugin version: `v2`

>   In response to procedure call `Status`, a compatible KMS plugin should return `v2` as `StatusResponse.version`,
>   "ok" as `StatusResponse.healthz` and a `key_id` (remote KMS KEK ID) as `StatusResponse.key_id`.

Reviewer: ```suggestion
 In response to the `Status` remote procedure call, a compatible KMS plugin should return its KMS compatibility
 version as `StatusResponse.version`. That status response should also include
 `"ok"` as `StatusResponse.healthz` and a `key_id` (remote KMS KEK ID) as `StatusResponse.key_id`.
 The Kubernetes project recommends you make your plugin
 compatible with the stable `v2` KMSAPI; Kubernetes {{< skew currentVersion >}} also supports the
 `v2beta1` API for KMS, and future Kubernetes releases are also likely to support that beta version.
```

##### KMS v2 {#developing-a-kms-plugin-gRPC-server-notes-kms-v2}

* KMS plugin version: `v2`

  In response to procedure call `Status`, a compatible KMS plugin should return `v2` as `StatusResponse.version`,
  "ok" as `StatusResponse.healthz` and a `key_id` (remote KMS KEK ID) as `StatusResponse.key_id`.

> 

Reviewer: Should we mention that there was a v2beta1 API? People might have a plugin that supports this API.

Also, are we announcing a deprecation period for that beta API (we probably should, unless it's deprecated-but-permanently-supported)?

Reviewer 2: They are both exactly the same and both will be supported forever as there is no value in breaking plugins.  How should we document that?

Reviewer: See https://github.com/kubernetes/website/pull/43398#pullrequestreview-1737781121 for how I'd write it.
